// profanity-matcher/profanity_matcher.js
var fs = require('fs')

function Profanity_Matcher() {

    this.badwords = fs.readFileSync(__dirname + '/naughty.csv').toString().split(",")

    this.scan = function(text, done) {
        // remove meaningless chars and pad with spaces to make detection better
        text.replace(/[-_=\+\/\\\.<>\?\*;\(\)\{\}:]/g, ' ')
        text = ' ' + text + ' '

        var profane = []
        this.badwords.forEach(function(word) {
            if (~text.indexOf(' ' + word + ' ')) {
                profane.push(word);
            }
        })

        return profane
    }

    this.add_word = function(word) {
        this.badwords.push(word)
        return save_words(this.badwords)
    }

    this.remove_word = function(word) {
        var index = this.badwords.indexOf(word)
        if (index != -1) {
            this.badwords.splice(index, 1)
            return save_words(this.badwords)
        } else {
            return false
        }
    }

    function save_words(badwords) {
        try {
            fs.writeFileSync(__dirname + '/naughty.csv', badwords.join())
            return true
        } catch (e) {
            return false
        }
    }

}

module.exports = Profanity_Matcher

if (require.main == module) {

    var pf = new Profanity_Matcher()
    var assert = require('assert')
    assert(pf.scan('hello there!') == [], 'detects simple good language')
    assert(pf.scan('fuck off') == ['fuck'], 'detects simple bad language')

}
